### ISSUE TITLE
Hello, I noticed that you used RPNloss in both the source and target domain images. Isn't this supervised? The method is no longer unsupervised.

### ISSUE THREAD

**Issue by duang33** [GiovanniPasq/DA-Faster-RCNN#21]
Hello, I noticed that you used RPNloss in both the source and target domain images. Isn't this supervised? The method is no longer unsupervised.

**Comment by GiovanniPasq** [comment]
Hi,  
I’ve double-checked the code and found no issues for the following reasons:

1. The target dataset is supposed to be unlabeled.  
2. Even if the target dataset were labeled, if you look at the code and how the loss is computed during training, you’ll see that it is not combined with the classification loss from the source dataset. Specifically, only the discriminator losses (`loss_image_d` and `loss_instance_d`) are taken from the target domain, averaged, and then summed into the total loss:

```python
loss_dict = model(data_source, False, 0.5)
loss_dict_target = model(data_target, True, 0.5)

loss_dict["loss_image_d"] += loss_dict_target["loss_image_d"]
loss_dict["loss_instance_d"] += loss_dict_target["loss_instance_d"]

loss_dict["loss_image_d"] *= 0.5
loss_dict["loss_instance_d"] *= 0.5

losses = sum(loss_dict.values())
```


As you can see, any other loss from the target is not included in the final losses value. Therefore, even if it were computed, it would be ignored during optimization.

Best regards

**Comment by duang33** [comment]
Got it, thank you.

**Comment by duang33** [comment]
> Hi, I’ve double-checked the code and found no issues for the following reasons:
> 
> 1. The target dataset is supposed to be unlabeled.
> 2. Even if the target dataset were labeled, if you look at the code and how the loss is computed during training, you’ll see that it is not combined with the classification loss from the source dataset. Specifically, only the discriminator losses (`loss_image_d` and `loss_instance_d`) are taken from the target domain, averaged, and then summed into the total loss:
> 
> loss_dict = model(data_source, False, 0.5)
> loss_dict_target = model(data_target, True, 0.5)
> 
> loss_dict["loss_image_d"] += loss_dict_target["loss_image_d"]
> loss_dict["loss_instance_d"] += loss_dict_target["loss_instance_d"]
> 
> loss_dict["loss_image_d"] *= 0.5
> loss_dict["loss_instance_d"] *= 0.5
> 
> losses = sum(loss_dict.values())
> As you can see, any other loss from the target is not included in the final losses value. Therefore, even if it were computed, it would be ignored during optimization.
> 
> Best regards

The code "proposals = self.label_and_sample_proposals(proposals, targets)" appears in both the source domain and the target domain, and finally proposals are indirectly passed into the code "loss_discriminatorProposal = self.discriminatorProposalDC5(box_features, target_domain, alpha)". Doesn't this mean that the target domain becomes supervised?

**Comment by GiovanniPasq** [comment]
Hi,
The discriminator does not use targets but rather target_domain, which is a boolean variable indicating whether the input image comes from the target or source domain. This information is known during training, since you know which data samples are from the target or source dataset. The code below shows how, in the case of target domain data, only the discriminator loss is computed:

```
loss_discriminatorProposal = self.discriminatorProposalDC5(box_features, target_domain, alpha)
predictions = self.box_predictor(box_features.mean(dim=[2, 3]))

if self.training:
    del features
    if target_domain:
        return [], loss_discriminatorProposal
    else:
        losses = self.box_predictor.losses(predictions, proposals)
        losses.update(loss_discriminatorProposal)
    if self.mask_on:
        proposals, fg_selection_masks = select_foreground_proposals(
            proposals, self.num_classes
        )
        # Since the ROI feature transform is shared between boxes and masks,
        # we don't need to recompute features. The mask loss is only defined
        # on foreground proposals, so we need to select out the foreground
        # features.
        mask_features = box_features[torch.cat(fg_selection_masks, dim=0)]
        del box_features
        losses.update(self.mask_head(mask_features, proposals))
    return [], losses
```

**Comment by duang33** [comment]
However, the box_features feature is extracted from the proposal variable, and the proposal variable adds the position of the GTbox

**Comment by GiovanniPasq** [comment]
The issue occurs when the target data passed is labeled — in that case, gt_instances is not None, which causes problems. If the target is unlabeled, everything works as expected.

To address this, could you please try replacing lines 181 and 182 of rcnn.py with the snippet below and test the network behavior?

```
if self.proposal_generator is not None:
    if target_domain:
        gt_instances = None
    proposals, proposal_losses = self.proposal_generator(images, features, gt_instances)
```

Let me know if this change improves the results or if any issues come up during testing. Thanks!

**Comment by duang33** [comment]
> The issue occurs when the target data passed is labeled — in that case, gt_instances is not None, which causes problems. If the target is unlabeled, everything works as expected.
> 
> To address this, could you please try replacing lines 181 and 182 of rcnn.py with the snippet below and test the network behavior?
> 
> ```
> if self.proposal_generator is not None:
>     if target_domain:
>         gt_instances = None
>     proposals, proposal_losses = self.proposal_generator(images, features, gt_instances)
> ```
> 
> Let me know if this change improves the results or if any issues come up during testing. Thanks!

Although this change is possible, it will cause a chain reaction. For example, GT must be input when training RPN, such as "assert gt_instances is not None, "RPN requires gt_instances in training!"" in the code.

**Comment by GiovanniPasq** [comment]
Yes, I checked rpn.py, and it does require gt_instances during training.
Another potential solution that comes to mind is to modify rpn.py to also accept the target_domain flag, and force it to follow the else branch of the if self.training condition when the input comes from the target dataset.
This way, you can bypass the need for gt_instances when processing target data.

**Comment by duang33** [comment]
> Yes, I checked rpn.py, and it does require gt_instances during training. Another potential solution that comes to mind is to modify rpn.py to also accept the target_domain flag, and force it to follow the else branch of the if self.training condition when the input comes from the target dataset. This way, you can bypass the need for gt_instances when processing target data.

Hello, after I modified it to my data, loss_instance_d: 1.793e-05
will always be very small, and the loss will not converge. Do you know why?

**Comment by GiovanniPasq** [comment]
You could try increasing the value of alpha (from 0.5 to 1) and modifying the discriminator architecture by changing the number of layers.
Unfortunately, depending on the dataset, some fine-tuning of these parameters is often necessary — as also highlighted by the authors of the original GRL paper, who used different discriminator configurations depending on the dataset.
See page 10 of the paper: https://arxiv.org/pdf/1409.7495

**Comment by duang33** [comment]
> You could try increasing the value of alpha (from 0.5 to 1) and modifying the discriminator architecture by changing the number of layers. Unfortunately, depending on the dataset, some fine-tuning of these parameters is often necessary — as also highlighted by the authors of the original GRL paper, who used different discriminator configurations depending on the dataset. See page 10 of the paper: https://arxiv.org/pdf/1409.7495

OK, thanks

**Comment by GiovanniPasq** [comment]
Hi,
I’ve made several updates to the architecture, specifically:
	•	Applied the fix in rpn.py
	•	Redesigned the discriminator architectures for both instance-level and image-level domain adaptation

Best regards,
Giovanni
